README.md
# Data analysis
- Document here the project: stockanalysis
- Description: Project Description
- Data Source:
- Type of analysis:

Please document the project the better you can.

# Startup the project

The initial setup.

Create virtualenv and install the project:
```bash
sudo apt-get install virtualenv python-pip python-dev
deactivate; virtualenv ~/venv ; source ~/venv/bin/activate ;\
    pip install pip -U; pip install -r requirements.txt
```

Unittest test:
```bash
make clean install test
```

Check for stockanalysis in gitlab.com/{group}.
If your project is not set please add it:

- Create a new project on `gitlab.com/{group}/stockanalysis`
- Then populate it:

```bash
##   e.g. if group is "{group}" and project_name is "stockanalysis"
git remote add origin git@github.com:{group}/stockanalysis.git
git push -u origin master
git push -u origin --tags
```

Functionnal test with a script:

```bash
cd
mkdir tmp
cd tmp
stockanalysis-run
```

# Install

Go to `https://github.com/{group}/stockanalysis` to see the project, manage issues,
setup you ssh public key, ...

Create a python3 virtualenv and activate it:

```bash
sudo apt-get install virtualenv python-pip python-dev
deactivate; virtualenv -ppython3 ~/venv ; source ~/venv/bin/activate
```

Clone the project and install it:

```bash
git clone git@github.com:{group}/stockanalysis.git
cd stockanalysis
pip install -r requirements.txt
make clean install test                # install and test
```
Functionnal test with a script:

```bash
cd
mkdir tmp
cd tmp
stockanalysis-run
```


notebooks/news
,Date,Title,Link,Ticker
0,2022-02-08 15:45:00,Taking On Volatility With Geared ETFs,https://finance.yahoo.com/news/taking-volatility-geared-etfs-204500574.html,SOXL
1,2022-02-02 00:00:00,ETFs End Jan. With $37B Inflows,https://finance.yahoo.com/news/etfs-end-jan-37b-inflows-050000994.html,SOXL
2,2022-01-17 19:07:00,The Truth Behind Trading Semiconductor (Chip) Stocks,https://www.investopedia.com/articles/active-trading/031815/truth-behind-trading-semiconductor-chip-stocks.asp?utm_campaign=quote-yahoo&utm_source=yahoo&utm_medium=referral,SOXL
3,2021-06-21 16:30:00,Direxion Changes Index for Semiconductor ETFs,https://finance.yahoo.com/news/direxion-changes-index-semiconductor-etfs-203000075.html,SOXL
4,2021-03-09 10:00:00,ETF Odds & Ends: IAU Share Split & More,https://finance.yahoo.com/news/etf-odds-ends-iau-share-150000904.html,SOXL
5,2021-02-08 09:30:00,ETF Odds & Ends: FANG+ ETNs Split,https://finance.yahoo.com/news/etf-odds-ends-fang-etns-143000923.html,SOXL
6,2020-07-10 11:04:00,"Chip stocks in focus, as NVDA surpasses INTC",https://finance.yahoo.com/video/chip-stocks-focus-nvda-surpasses-150450559.html,SOXL
7,2020-06-29 10:45:00,Best Leveraged/Inverse ETFs of June,https://finance.yahoo.com/news/best-leveraged-inverse-etfs-june-144502308.html,SOXL
8,2020-05-11 10:27:00,6 Leveraged ETFs Up At Least 25% Last Week,https://finance.yahoo.com/news/6-leveraged-etfs-least-25-142702621.html,SOXL
9,2020-04-21 16:00:00,Bullish ETF Strategies to Capitalize on Tech World,https://finance.yahoo.com/news/bullish-etf-strategies-capitalize-tech-200000061.html,SOXL
10,2020-04-14 10:15:00,Semiconductor stocks are a bellwether and confirm the market's view of growth: Prudential Financial Chief Market Strategist,https://finance.yahoo.com/video/semiconductor-stocks-bellwether-confirm-markets-141534320.html,SOXL
11,2020-04-07 00:00:00,Best Performing ETFs Of The Year,https://finance.yahoo.com/news/best-performing-etfs-050000748.html,SOXL
12,2020-03-12 04:52:00,"Trump announces coronavirus travel ban on Europe, Dow futures plummet",https://finance.yahoo.com/news/trump-announces-coronavirus-travel-ban-on-europe-dow-futures-plummet-085217749.html,SOXL
13,2020-02-28 09:05:00,Consider Quick Upside From Downside in Chips With This ETF,https://finance.yahoo.com/news/consider-quick-upside-downside-chips-140540062.html,SOXL
14,2020-02-17 08:34:00,4 Best Leveraged ETFs of Last Week,https://finance.yahoo.com/news/4-best-leveraged-etfs-last-130001169.html,SOXL
15,2019-12-31 11:00:00,9 Leveraged ETFs That More Than Doubled in 2019,https://finance.yahoo.com/news/9-leveraged-etfs-more-doubled-160004197.html,SOXL
16,2019-12-22 23:31:00,3 ETFs to Capitalize on Semiconductor Strength the Past Decade,https://finance.yahoo.com/news/3-etfs-capitalize-semiconductor-strength-043147948.html,SOXL
17,2019-12-18 08:00:00,Wall Street to See Best Year Since 2013: Will ETF Rally Last?,https://finance.yahoo.com/news/wall-street-see-best-since-130001182.html,SOXL
18,2019-12-16 18:44:00,ETF Winners as US-China Deal Eases Trade Tensions,https://finance.yahoo.com/news/etf-winners-us-china-deal-234411052.html,SOXL
19,2019-12-16 16:13:00,"Stock market news: December 16, 2019",https://finance.yahoo.com/news/stock-market-news-december-16-2019-131306796.html,SOXL
20,2019-12-03 16:19:00,"Stock market news: December 3, 2019",https://finance.yahoo.com/news/stock-market-news-december-3-2019-130245185.html,SOXL
21,2019-11-20 10:10:00,Leveraged ETFs That Are Up 25% Plus at Halfway Q4,https://finance.yahoo.com/news/leveraged-etfs-25-plus-halfway-151003759.html,SOXL
22,2019-11-08 15:49:00,Top Performing Leveraged ETFs,https://finance.yahoo.com/news/top-performing-leveraged-etfs-204908779.html,SOXL
23,2019-11-06 10:10:00,Why You Should Bet on Leveraged ETFs in November,https://finance.yahoo.com/news/why-bet-leveraged-etfs-november-151003441.html,SOXL
24,2019-10-28 18:14:00,ETFs to Gain as Sino-US Trade Talks Progress,https://finance.yahoo.com/news/etfs-gain-sino-us-trade-221410569.html,SOXL
25,2019-10-14 16:59:00,ETF Winners as 'Phase 1' Deal Eases Trade Tensions,https://finance.yahoo.com/news/etf-winners-phase-1-deal-205908024.html,SOXL
26,2019-10-11 17:30:00,ETF Areas to Shine on 'Very Good' Trade Negotiation,https://finance.yahoo.com/news/etf-areas-shine-very-good-213009045.html,SOXL
27,2019-10-10 17:45:00,Chip ETFs in Focus as Trump Blacklists More Chinese Tech Firms,https://finance.yahoo.com/news/chip-etfs-focus-trump-blacklists-214509682.html,SOXL
28,2019-10-07 10:49:00,Can Semiconductors Dodge the Trade War Crossfire?,https://finance.yahoo.com/news/semiconductors-dodge-trade-war-crossfire-144941311.html,SOXL
29,2019-09-28 15:01:00,Huawei Skips US Parts in 5G Base Stations: Chip ETFs at Risk?,https://finance.yahoo.com/news/huawei-skips-us-parts-5g-190107457.html,SOXL
30,2019-09-26 13:00:00,4 Leveraged ETF Areas That Are Up 100% This Year,https://finance.yahoo.com/news/4-leveraged-etf-areas-100-170005812.html,SOXL
31,2019-09-26 09:56:00,Leveraged Semiconductor ETFs Are Back In The Spotlight,https://finance.yahoo.com/news/leveraged-semiconductor-etfs-back-spotlight-135639131.html,SOXL
32,2019-09-21 10:45:00,ETFs in Focus as Deputy-Level U.S.-China Trade Talks Begin,https://finance.yahoo.com/news/etfs-focus-deputy-level-u-144502618.html,SOXL
33,2019-09-16 14:48:00,ETF Winners as Sino-US Trade War Tensions Ebb,https://finance.yahoo.com/news/etf-winners-sino-us-trade-184806517.html,SOXL
34,2019-09-11 11:30:00,8 High-Flying Leveraged ETFs YTD,https://finance.yahoo.com/news/8-high-flying-leveraged-etfs-153003898.html,SOXL
35,2019-09-09 08:00:00,Top and Flop Leveraged ETFs of Last Week,https://finance.yahoo.com/news/top-flop-leveraged-etfs-last-120012870.html,SOXL
36,2019-09-02 10:00:00,8 Top-Performing Leveraged ETFs of Last Week,https://finance.yahoo.com/news/8-top-performing-leveraged-etfs-140002663.html,SOXL

stockanalysis/data.py
import pandas as pd
import yfinance as yf
import numpy as np
import pandas_ta as pta
import tweepy
import requests
from bs4 import BeautifulSoup
from stockanalysis.utils import *
from pandas_datareader import data as pdr
yf.pdr_override()

def get_technical(symbol="INFY.NS",start="2017-01-01", end="2021-04-30"):
    '''returns a DataFrame with stock technical data'''
    df = pdr.get_data_yahoo(symbol, start=start, end=end)
    df.drop(columns=['Adj Close'],inplace=True)

    df['ema12'] = get_ema(df, column='Close', period=12)
    df['ema21'] = get_ema(df, column='Close', period=21)
    df['ema26'] = get_ema(df, column='Close', period=26)
    df['ema34'] = get_ema(df, column='Close', period=34)
    df['ema55'] = get_ema(df, column='Close', period=55)
    df['ema99'] = get_ema(df, column='Close', period=99)
    df['ema200'] = get_ema(df, column='Close', period=200)
    df['hma12'] = get_hma(df, column='Close', period=12)
    df['hma21'] = get_hma(df, column='Close', period=21)
    df['hma26'] = get_hma(df, column='Close', period=26)
    df['hma34'] = get_hma(df, column='Close', period=34)
    df['hma55'] = get_hma(df, column='Close', period=55)
    df['hma99'] = get_hma(df, column='Close', period=99)
    df['hma200'] = get_hma(df, column='Close', period=200)
    df['rsi'] = get_rsi(df, period=14)
    df['atr'] = get_atr(df, period=14)
    df['bb_upper'] = get_bband(df, period=20, std=2)['BBU_20_2.0']
    df['bb_lower'] = get_bband(df, period=20, std=2)['BBL_20_2.0']
    df['macd_signal'] = get_macd(df, fast=12, slow=26,
                                 signal=9)['MACD_12_26_9']
    df['macd_line'] = get_macd(df, fast=12, slow=26, signal=9)['MACDs_12_26_9']
    df['adx'] = get_adx(df, length=14)['ADX_14']
    df['vwap'] = get_vwap(df)
    cleaned_df = clean_data(df)
    return cleaned_df

def get_fundamental():
    '''returns a DataFrame of stock fundamental data'''
    pass

def get_tweets():
    '''returns a DataFrame with tweets realted to stock'''


def get_stock_news():
    '''returns a DataFrame with latest news realted of stock'''
    pass

def get_recommendation():
    '''returns a DataFrame with stock analysis buy sell recommendations'''
    pass

def clean_data(df, test=False):
    '''returns a DataFrame without outliers and missing values'''
    df = df.dropna(how='any')
    df = df.reset_index()
    return df


if __name__ == '__main__':
    from stockanalysis.database import *
    import datetime
    from dateutil.relativedelta import relativedelta

    conn = connection.connect(**config)
    mycursor = conn.cursor(dictionary=True)
    query = f"""
    SELECT Date FROM stocksdb.raw_technical where Stock_id=1 ORDER BY Date DESC LIMIT 1;
    ;"""
    mycursor.execute(query)
    stock_db_lastdate = mycursor.fetchone()

    if stock_db_lastdate == None:
        end_date = datetime.datetime.now().strftime("%Y-%m-%d")
        start_date = (datetime.datetime.now() - datetime.timedelta(days=5 * 365)).strftime("%Y-%m-%d")
        df = get_technical(symbol="INFY.NS", start=start_date, end=end_date)
        print(df)
    else:
        date_time_str = stock_db_lastdate['Date'].split()[0]
        date_time_obj = datetime.strptime(date_time_str, '%Y/%m/%d')
        end_date = datetime.datetime.now().strftime("%Y-%m-%d")
        start_date = date_time_obj + datetime.timedelta(days=1)
        df = get_technical(symbol="INFY.NS", start="2017-01-01", end="2022-04-30")
        print(df)

    #print(stock_db_lastdate['Date'].split()[0])
    #df = get_technical()
    #print(df)
    #cleaned_df = cleaned_df
    #cleaned_df.to_csv("raw_technical.csv", sep='\t', encoding='utf-8')
    #query = ("INSERT INTO stocksdb.TechnicalData (Date,StockID) "
    #"VALUES (%s, 1)")
    #db.SaveDFToTable(query, cleaned_df)
